ipyparallel/cluster/shellcmd.py

```diff
@@ -0,0 +1,860 @@
+            userdir = os.environ["USERPROFILE"]
+        elif platform == Platform.Linux:
+            userdir = os.environ["HOME"] if "HOME" in os.environ else ""
+        self.filename = filename.replace(
+            "${userdir}", userdir
+        )  # replace possible ${userdir} placeholder
+
+    def _output_msg(self, level, msg):
+        with open(self.filename, "a") as f:
+            f.write(f"{level:8} {datetime.now()} {msg}\n")
+
+    def info(self, msg):
+        self._output_msg("info", msg)
+
+    def warning(self, msg):
+        self._output_msg("warning", msg)
+
+    def error(self, msg):
+        self._output_msg("error", msg)
+
+    def debug(self, msg):
+        self._output_msg("debug", msg)
+
+
+class ShellCommandReceive:
+    """
+    Helper class for receiving and performing shell commands in a platform independent form
+
+    All supported shell commands have a cmd_ prefix. When adding new functions make sure that the is an
+    equivalent in the ShellCommandSend class. When a command failed a non zero exit code will be returned.
+    Hence, the ShellCommandSend class always uses subprocess.check_output for assessing if the command was
+    successfull. Some command require information to be returned (cmd_exists, cmd_start, cmd_running) which
+    is written to stdout in the following form: __<key>=<value>__
+    """
+
+    def _log(self, msg):
+        if not self.log:
+            return
+        self.log.info(f"[id={self.ranid}] {msg}")
+
+    def __init__(self, debugging=False, use_break_way=True, log=None):
+        self.debugging = debugging
+        self.log = None
+        self.platform = Platform.get()
+        self.use_break_way = use_break_way
+        if log:
+            if isinstance(log, str):
+                self.log = SimpleLog(log)
+            else:
+                self.log = log
+        elif "SHELLCMD_LOG" in os.environ:
+            self.log = SimpleLog(os.environ["SHELLCMD_LOG"])
+
+        self.ranid = None
+        if self.log:
+            self.ranid = randint(0, 999)
+            self._log("ShellCommandReceive instance created")
+
+    def __del__(self):
+        if self.log:
+            self._log("ShellCommandReceive instance deleted")
+
+    def _linux_quote(self, p):
+        if "'" in p:
+            return '"' + p + '"'
+        else:
+            return p
+
+    def cmd_start(self, start_cmd, env=None, output_file=None):
+        if env:
+            # add provided entries to environment
+            if isinstance(env, str):
+                if env[0] == '"' and env[-1] == '"':
+                    env = env.strip('"')  # occurs under windows cmd
+                if "\\'" in env:
+                    env = env.replace("\\'", "'")  # replace quoted
+                env_dict = eval(env)
+                assert isinstance(env_dict, dict) is True
+                env = env_dict
+
+            assert isinstance(env, dict) is True  # make sure that env is a dictionary
```

Comment:
General tip: you don't need to compare booleans with True, you can `assert isinstance(env_dict, dict)`. But instead of that, we should actually raise a real error that can't be optimized-out.

```suggestion
           if not isinstance(env_dict, dict):
               raise TypeError(f"env must be a dict, got {env_dict!r}")
```

The identical assertion 3 lines above can be removed.

Reply:
right, fixed by commit b957d40